stetho/src/main/java/com/facebook/stetho/inspector/elements/Document.java

```diff
@@ -0,0 +1,602 @@
+  }
+
+  public int getNodeIdForElement(Object element) {
+    verifyThreadAccess();
+    return mObjectIdMapper.getIdForObject(element);
+  }
+
+  public Object getElementForNodeId(int id) {
+    verifyThreadAccess();
+    return mObjectIdMapper.getObjectForId(id);
+  }
+
+  public DocumentView getDocumentView() {
+    verifyThreadAccess();
+    return mShadowDOM;
+  }
+
+  public Object getRootElement() {
+    verifyThreadAccess();
+
+    Object rootElement = mDocumentProvider.getRootElement();
+    if (rootElement == null) {
+      // null for rootElement is not allowed. We could support it, but our current
+      // implementation won't ever run into this, so let's punt on it for now.
+      throw new IllegalStateException();
+    }
+
+    if (mShadowDOM == null) {
+      mShadowDOM = new ShadowDocument(mDocumentProvider.getRootElement());
+      createShadowDOMUpdate().commit();
+      mDocumentProvider.setListener(new ProviderListener());
+    } else if (rootElement != mShadowDOM.getRootElement()) {
+      // We don't support changing the root element. This is handled differently by the
+      // protocol than updates to an existing DOM, and we don't have any case in our
+      // current implementation that causes this to happen, so let's punt on it for now.
+      throw new IllegalStateException();
+    }
+
+    return rootElement;
+  }
+
+  public void findMatchingElements(String query, Accumulator<Integer> matchedIds) {
+    verifyThreadAccess();
+
+    final Pattern queryPattern = Pattern.compile(Pattern.quote(query), Pattern.CASE_INSENSITIVE);
+    final Object rootElement = mDocumentProvider.getRootElement();
+
+    findMatches(queryPattern, rootElement, matchedIds);
+  }
+
+  private void findMatches(Pattern queryPattern, Object element, Accumulator<Integer> matchedIds) {
+    final ElementInfo info = mShadowDOM.getElementInfo(element);
+
+    for (int i = 0, size = info.children.size(); i < size; i++) {
+      final Object childElement = info.children.get(i);
+
+      if (doesElementMatch(queryPattern, childElement)) {
+        matchedIds.store(mObjectIdMapper.getIdForObject(childElement));
+      }
+
+      findMatches(queryPattern, childElement, matchedIds);
+    }
+  }
+
+  private boolean doesElementMatch(Pattern queryPattern, Object element) {
+    AttributeListAccumulator accumulator = new AttributeListAccumulator();
+    NodeDescriptor descriptor = mDocumentProvider.getNodeDescriptor(element);
+
+    descriptor.getAttributes(element, accumulator);
+
+    for (int i = 0, N = accumulator.size(); i < N; i++) {
+      if (queryPattern.matcher(accumulator.get(i)).find()) {
+        return true;
+      }
+    }
+
+    return queryPattern.matcher(descriptor.getNodeName(element)).find();
+  }
+
+  private ChildEventingList acquireChildEventingList(
+      Object parentElement,
+      DocumentView documentView) {
+    ChildEventingList childEventingList = mCachedChildEventingList;
```

Comment:
This method has a bug in it. We need to set `mCachedChildEventingList` to null! Otherwise we could hand it out to multiple call sites in the event of some recursive algorithm or reentrant code block.